src/Elastic.Apm/Metrics/MetricsCollector.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Elastic.Apm.Api;
+using Elastic.Apm.Config;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Metrics.MetricsProvider;
+using Elastic.Apm.Report;
+using Timer = System.Timers.Timer;
+
+namespace Elastic.Apm.Metrics
+{
+	/// <summary>
+	/// Iterates through a list of <see cref="IMetricsProvider" /> and
+	/// sends the values through an <see cref="IPayloadSender" /> instance.
+	/// It collects the metrics on an interval read from the <see cref="IConfigurationReader" /> which is
+	/// passed to the constructor.
+	/// In case reading a value from an <see cref="IMetricsProvider" /> fails it retries <see cref="MaxTryWithoutSuccess" />
+	/// times,
+	/// after that it prints a log and won't retry anymore. This is to avoid endlessly trying to read values without success.
+	/// </summary>
+	internal class MetricsCollector : IDisposable, IMetricsCollector
+	{
+		internal const int MaxTryWithoutSuccess = 5;
+
+		private static int _syncPoint;
+
+		/// <summary>
+		/// List of all providers that can provide metrics values.
+		/// Add new providers to this list in case you want the agent to collect more metrics
+		/// </summary>
+		internal readonly List<IMetricsProvider> MetricsProviders;
+
+		private readonly IApmLogger _logger;
+
+		private readonly IPayloadSender _payloadSender;
+
+		private readonly Timer _timer;
+
+		public MetricsCollector(IApmLogger logger, IPayloadSender payloadSender, IConfigurationReader configurationReader)
+		{
+			_logger = logger.Scoped(nameof(MetricsCollector));
+
+			MetricsProviders = new List<IMetricsProvider>
+			{
+				new FreeAndTotalMemoryProvider(),
+				new ProcessWorkingSetAndVirtualMemoryProvider(),
+				new SystemTotalCpuProvider(_logger),
+				new ProcessTotalCpuTimeProvider()
+			};
+
+			_payloadSender = payloadSender;
+
+			var interval = configurationReader.MetricsIntervalInMillisecond;
+
+			// ReSharper disable once CompareOfFloatsByEqualityOperator
+			if (interval == 0)
+			{
+				_logger.Info()?.Log("Collecting metrics is disabled - the agent won't collect metrics");
+				return;
+			}
+
+			_logger.Info()?.Log("Collecting metrics in {interval} milliseconds interval", interval);
+			_timer = new Timer(interval);
+			_timer.Elapsed += (sender, args) => { CollectAllMetrics(); };
+		}
+
+		public void StartCollecting() => _timer?.Start();
+
+		internal void CollectAllMetrics()
```

Comment:
We should prevent exception escaping from timer callback.

Reply:
Don't we already do that?

The only part outside the try-catch is this:
```	
var sync = Interlocked.CompareExchange(ref _syncPoint, 1, 0);
if (sync != 0) return;
``` 
And according to the docu `CompareExchange` only throws if the 1. parameter is `null`, and I'm very sure that's not the case here. Do you want me to move that into the try block?

Reply:
I agree - it's handled already. I would have extracted that and thread safety to a separate method and placing the business logic in `CollectAllMetricsInternal` but that is a matter of taste.